test/UnitWOPIHttpRedirectLoop.cpp

```diff
@@ -52,22 +53,25 @@ class UnitWopiHttpRedirectLoop : public WopiTestServer
 
             assertCheckFileInfoRequest(request);
 
-            LOK_ASSERT_MESSAGE("It is expected to stop requesting after 21 redirections", fileId <= 22);
+            std::string sExpectedMessage = "It is expected to stop requesting after " + std::to_string(RedirectionLimit) + " redirections";
+            LOK_ASSERT_MESSAGE(sExpectedMessage, fileId <= RedirectionLimit + 1);
```

Comment:
Shouldn't we expect that the **last** retry be RedirectionLimit, and no redirections after that? Meaning, should we check that `fileId <= RedirectionLimit` ?

Reply:
we start with `fileId = 1;` (see line 44, above) not from 0 which is the "initial URL". So `RedirectionLimit + 1` is good.

/wopi/files/0    fileId = 1     redirectionLimit=21
/wopi/files/1    fileId = 2     redirectionLimit=20
...
/wopi/files/20  fileId = 21   redirectionLimit=1
/wopi/files/21  fileId = 22   redirectionLimit=0
/wopi/files/22 is not requested

I used fileId because it's easier than extracting X from /wopi/files/X

Reply:
Right. That's the issue (granted, it's a minor one). You see, the limit is called 'Redirection Limit', but it's not representing the number of redirections. It makes more sense to start from 0 (meaning, we didn't do any redirections). Then, we ++ and it becomes 1 *after* we get the first request and reply back with the *first* redirection.

Yes, this is the server side, and we are counting the number of times the server returned 302. We are *not* counting the number of times the client requested the same URL (of course there is a relationship between the request and the response, but we are counting the response and not the requests).

When we do the assertion, at that point, we can compare the number of requests, or the number of redirections, or any combination we want. But the numbers would represent exactly what they are counting (the number of times we reply).

The reason I'm mentioning all this is because when there is a +1 when comparing with a counter, it typically implies that it has the wrong name (in this case, it seems we are really counting the requests, but comparing it with the redirections, which is unnecessary and confusing). This means that it will be confusing to someone else (even if it works correctly). 

Since we are doing both the count and the assertion in the same component (the server, when handling the request), I expect us to count the number of times we did reply with 302 (i.e. did redirection) and refuse to do any more when we exceed the limit. That is, we start with 0 (no redirections done), compare `fileId < RedirectionLimit` (the last time this assertion passes we will send the 21st redirection and fileId will be 21), and we expect *not* to get the 22nd request. If we get another one, fileId will be 21 and that is RedirectionLimit. We fail the assertion. (I would also rename fileId to redirectionCount, then it becomes very clear: redirectionCount = 0; and we assert `redirectionCount < RedirectionLimit`, then every time we send back 302 we ++).